pkg/virt-operator/resource/apply/apps.go

```diff
@@ -78,7 +103,124 @@ func (r *Reconciler) syncDeployment(deployment *appsv1.Deployment) error {
 	return nil
 }
 
-func (r *Reconciler) syncDaemonSet(daemonSet *appsv1.DaemonSet) error {
+func (r *Reconciler) isDaemonSetUpToDate(daemonSet *appsv1.DaemonSet) bool {
+	imageTag, imageRegistry, id := getTargetVersionRegistryID(r.kv)
+
+	return daemonSet.Annotations[v1.InstallStrategyVersionAnnotation] == imageTag &&
+		daemonSet.Annotations[v1.InstallStrategyRegistryAnnotation] == imageRegistry &&
+		daemonSet.Annotations[v1.InstallStrategyIdentifierAnnotation] == id
+}
+
+func setMaxUnavailable(daemonSet *appsv1.DaemonSet, maxUnavailable intstr.IntOrString) {
+	daemonSet.Spec.UpdateStrategy.RollingUpdate = &appsv1.RollingUpdateDaemonSet{
+		MaxUnavailable: &maxUnavailable,
+	}
+}
+
+func generateDaemonSetPatch(oldDs, newDs *appsv1.DaemonSet) ([]byte, error) {
+	newSpec, err := json.Marshal(newDs.Spec)
+	if err != nil {
+		return nil, err
+	}
+
+	ops, err := getPatchWithObjectMetaAndSpec([]string{
+		fmt.Sprintf(testGenerationJSONPatchTemplate, oldDs.ObjectMeta.Generation),
+	}, &newDs.ObjectMeta, newSpec)
+	if err != nil {
+		return nil, err
+	}
+	return generatePatchBytes(ops), nil
+}
+
+func isDaemonSetReady(daemonSet *appsv1.DaemonSet) bool {
+	return daemonSet.Status.DesiredNumberScheduled == daemonSet.Status.NumberReady &&
+		daemonSet.Status.NumberReady == daemonSet.Status.UpdatedNumberScheduled
+}
+
+func (r *Reconciler) patchDaemonSet(oldDs, newDs *appsv1.DaemonSet) (*appsv1.DaemonSet, error) {
+	patch, err := generateDaemonSetPatch(oldDs, newDs)
+	if err != nil {
+		return nil, err
+	}
+
+	newDs, err = r.clientset.AppsV1().DaemonSets(r.kv.Namespace).Patch(
+		context.Background(),
+		newDs.Name,
+		types.JSONPatchType,
+		patch,
+		metav1.PatchOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("unable to update daemonset %+v: %v", oldDs, err)
+	}
+	return newDs, nil
+}
+
+func (r *Reconciler) processCanaryUpgrade(cachedDaemonSet, newDS *appsv1.DaemonSet) (bool, error, CanaryUpgradeStatus) {
```

Comment:
it seems like this should be as simple as

```
if updatedAndReadyPods == 0 {
    set maxUnavailable = 1
} else if updatedAndReadyPods > 0 && updatedAndReadyPods < totalExpected {
   Set maxUnavailable = 10%
} else if updatedAndReadyPods == totalExpected {
  Set maxUnavailable = 1%
}
```

The number of `updatedAndReadyPods` can be calculated by looking at virt-handler pods in the pod cache.